accord-core/src/main/java/accord/impl/SafeCommandsForKey.java

```diff
@@ -165,22 +167,30 @@ public <D> CommandsForKey listenerUpdate(Command command)
     }
 
     @VisibleForImplementation
-    public <D> CommandsForKey updateLastExecutionTimestamps(Timestamp executeAt, boolean isForWriteTxn)
+    public <D> CommandsForKey updateLastExecutionTimestamps(SafeCommandStore safeStore, Timestamp executeAt, boolean isForWriteTxn)
     {
         CommandsForKey current = current();
 
         Timestamp lastWrite = current.lastWriteTimestamp();
 
         if (executeAt.compareTo(lastWrite) < 0)
+        {
+            if (!safeStore.commandStore().safeToReadAt(executeAt).contains(key.toUnseekable()))
```

Comment:
`bootstrappedAt` and `safeToRead` are distinct but related concepts - `bootstrappedAt` defines the `txnId` bounds we expect to maintain data for locally; `safeToRead` defines `executeAt` bounds we can safely participate in transaction execution for. `safeToRead` is defined by the no-op transaction we execute after a bootstrap is initiated, and creates a global bound before which we know we have complete data from our bootstrap. There’s a smearing period during bootstrap where some keys may be ahead of others, for instance, since we do not create a precise instant in the transaction log for bootstrap so as to avoid impeding execution.

We can also bootstrap ranges at different times, and have a transaction that participates in both ranges - in this case one of the portions of the transaction may be totally unordered with respect to other transactions in that range because both occur prior to the `bootstrappedAt` point, so their dependencies are entirely erased. We can also re-bootstrap the same range because bootstrap failed, and leave dangling transactions to execute which then execute in an unordered fashion.

We can also update `safeToRead` when we go stale, to remove ranges we may have bootstrapped but that are now known to be incomplete. In this case we permit transactions to execute in any order for the unsafe key ranges. But they may still be ordered for other key ranges they participate in.

We could plausibly pro-actively erase transactions that only participate in stale ranges, but I’m not sure this would buy much. Perhaps at a later date.